src/server/client.js

```diff
@@ -0,0 +1,555 @@
+                .then(project => project.addCollaborator(username));
+        }
+    },
+
+    'permission-elevation-request': function(msg) {
+        if (this.hasProject()) {
+            const projectId = this.projectId;
+            return Projects.getRawProjectById(this.projectId)
+                .then(metadata => {
+                    const {owner} = metadata;
+                    const sockets = NetworkTopology.getSocketsAtProject(projectId);
+                    const owners = sockets.filter(socket => socket.username === owner);
+
+                    owners.forEach(socket => socket.send(msg));
+                });
+        }
+    },
+
+    'request-room-state': function() {
+        if (this.hasProject()) {
+            return NetworkTopology.getRoomState(this.projectId)
+                .then(msg => {
+                    msg.type = 'room-roles';
+                    this.send(msg);
+                });
+        }
+    },
+
+    ///////////// Import/Export /////////////
+    'export-room': function(msg) {
+        if (this.hasProject()) {
+            const projectId = this.projectId;
+            const occupantForRole = {};
+
+            // Get the first occupant for each role
+            NetworkTopology.getSocketsAtProject(projectId).reverse()
+                .forEach(socket => {
+                    occupantForRole[socket.roleId] = socket;
+                });
+
+            return Projects.getById(this.projectId)
+                .then(project => {
+                    // For each role...
+                    //   - if it is occupied, request the content
+                    //   - else, use the content from the database
+                    return project.getRoleIds()
+                        .then(ids => {
+                            const fetchers = ids.map(id => {
+                                const occupant = occupantForRole[id];
+                                if (occupant) {
+                                    return occupant.getProjectJson()
+                                        .catch(err => {
+                                            this._logger.info(`Failed to retrieve project via ws. Falling back to content from database... (${err.message})`);
+                                            return project.getRoleById(id);
+                                        });
+                                }
+                                return project.getRoleById(id);
+                            });
+
+                            return Q.all(fetchers);
+                        });
+                })
+                .then(roles => {
+                    const roleContents = roles.map(content =>
+                        Utils.xml.format('<role name="@">', content.ProjectName)
+                        + content.SourceCode + content.Media + '</role>'
+                    );
+                    return Utils.xml.format('<room name="@" app="@">', this.name, Utils.APP) +
```

Comment:
Where is this.name set?

Reply:
That's a mistake from the port. This function used to belong in `ActiveRoom` which has `this.name` set (matching the netsblox project's name). Since this has been moved to the client class, this needs to be looked up from the project instead. 

Good catch!